theano/gpuarray/opt.py

```diff
@@ -215,6 +256,247 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(NavigatorOptimizer):
+    """
+    Transfer the graph as a whole to GPU instead of transfering node by node.
+    """
+
+    def __init__(self, local_optimizers_all, local_optimizers_map):
+        self.local_optimizers_all = local_optimizers_all
+        self.local_optimizers_map = local_optimizers_map
+        self.failure_callback = None
+
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        time_opts = {}
+        node_created = {}
+        process_count = {}
+        t_topo = time.time()
+        topo = fgraph.toposort()
+        time_topo = time.time()
+        toposort_timing = time_topo - t_topo
+
+        # Building a new graph
+        # Iterating through inputs of graph
+        for i in fgraph.inputs:
+            # Do not move *int* scalar to the GPU.
+            target = getattr(i.tag, 'target', None)
+            if (target != 'cpu' and
+                    isinstance(i.type, tensor.TensorType) and
+                    (i.ndim > 0 or 'int' not in i.dtype)):
+                mapping[i] = as_gpuarray_variable(i, target)
```

Comment:
I get some problem with the context_name here. I get this traceback when i try to compile a theano function.(though this doesn't happen when i run the tests, only happens when i try it from jupyter-notebook) 
http://pastebin.com/qYK26Awv